File: observer.cc
#include <iostream>
#include <vector>
#include<algorithm>
class Observer {
public:
    virtual void update() = 0;
};

class Subject {
public:
    virtual void Attach(Observer* obs) = 0;
    virtual void Detach(Observer* obs) = 0;
    virtual void Notify() = 0;
    virtual void setStatus(int status) = 0;
    virtual int getStatus() = 0;
};

class OfficeDoc : public Subject {
private:
    std::vector<Observer*> myObs;
    std::string mySubjectName;
    int m_status;

public:
    OfficeDoc(std::string name) {
        mySubjectName = name;
        m_status = 0;
    }

    void Attach(Observer* obs) {
        myObs.push_back(obs);
    }

    void Detach(Observer* obs) {

        myObs.erase(std::remove(myObs.begin(), myObs.end(), obs), myObs.end());
        // auto it = std::find(myObs.begin(),myObs.end(), obs);
        // if (it != myObs.end()) {
        //     myObs.erase(it);
        // }
    }

    void Notify() {
        for (Observer* obs : myObs) {
            obs->update();
        }
    }

    void setStatus(int status) {
        m_status = status;
        std::cout << "SetStatus subject[" << mySubjectName << "] status:" << status << std::endl;
    }

    int getStatus() {
        return m_status;
    }
};

class DocExplorer : public Observer {
private:
    std::string myObsName;

public:
    DocExplorer(std::string name, Subject* sub) {
        myObsName = name;
        sub->Attach(this);
    }

    void update() {
        std::cout << "update observer[" << myObsName << "]" << std::endl;
    }
};

int main() {
    // std::cout << "Hello World" << std::endl;
    Subject* subjectA = new OfficeDoc("subject A");
    Observer* observerA = new DocExplorer("observer", subjectA);
    subjectA->setStatus(1);
    subjectA->Notify();
    // "update observer[observer]" is printed when subjectA->Notify(); is called,
    // and it invokes the update method of observerA.

    delete observerA;
    delete subjectA;

    return 0;
}

File: observer.java
import java.util.*;
interface Observer{public void update();}//public (1);
interface Subject{
    public void Attach(Observer obs);
    public void Detach(Observer obs);
    public void Notify();
    public void setStatus(int status);
    public int getStatus();
}
class OfficeDoc implements Subject{
    private List<Observer> myObs;//list<2>
    private String mySubjectName;
    private int m_status;
    public OfficeDoc(String name)
    {
        mySubjectName=name;
        this.myObs=new ArrayList<Observer>();
        m_status=0;
    }
    public void Attach(Observer obs){this.myObs.add(obs);}
    public void Detach(Observer obs){this.myObs.remove(obs);}
    public void Notify(){
        for(Observer obs:this.myObs)
        {
            obs.update();//(3)
        }
    }
    public void setStatus(int status)
    {
        m_status=status;
        System.out.println("SetStatus subject["+mySubjectName+
        "] status:"+status);
    }
     public int getStatus(){return m_status;}
    
}
class DocExplorer implements Observer{
    private String myObsName;
    public DocExplorer (String name,Subject sub)//,(4) sub
    {
        myObsName=name;
        sub.Attach(this);
    }
    public void update(){
        System.out.println("update observer["+myObsName+"]");
    }
   
}
class ObserverTest{
    public static void main(String[] args)
    {
        // System.out.println("Hello World");
        Subject subjectA=new OfficeDoc("subject A");
        Observer observerA=new DocExplorer("observer",subjectA);
        subjectA.setStatus(1);
        subjectA.Notify();
        // "update observer[observer]" is printed when subjectA.Notify(); /-
        // is called, and it invokes the update method of observerA.
       
    }

}